File: log_server/ws/message.go
package ws

type MessageType string

const (
	EXECUTE MessageType = "execute"
	ABORT   MessageType = "abort"
	INFO    MessageType = "info"
	ERROR   MessageType = "error"
	PAUSE   MessageType = "pause"
	RESUME  MessageType = "resume"
)

type Message struct {
	Type MessageType `json:"type"`
	Data any         `json:"data"`
}

File: web_client/src/Message.ts
type CPUStat = {
    user: number, // milliseconds
    system: number, // milliseconds
}

type Message = {
    type: "execute" | "info" | "error",
    data: string,
} | {
    type: "abort" | "pause" | "resume"
} | {
    type: "exec_finished",
    data: {
        duration: number, // milliseconds
        user_time: number, // milliseconds
        system_time: number, // milliseconds
        exit_status: number // milliseconds
    }
} | {
    type: "stats",
    data: {
        process: CPUStat,
        global: number, // milliseconds
        rss: number, // kB
        peak_rss: number, // kB
        delta: number // milliseconds
    }
} | {
    type: "fulfilled",
    data: Exclude<Message["type"], "fulfilled">
}

type MessageType = Message["type"]

export type { MessageType, Message }
